UCR Conditional Aggregations

esoergel opened the pull request:
This adds as-yet-undocumented support for aggregating rows in a UCR Report by SQL conditional expressions.  The target use-case is the ICDS Mobile UCR v2 migration, which would benefit greatly from a subset of this.

For ICDS, we need to be able to group by age-group (0-6 months, 6-12 months, etcetera).  I'm also considering a more targeted approach here, because as-implemented, this is maybe overly powerful.  However, I've heard of very similar use cases, which don't fit quite so narrow a bucket - Emord has one which would be an extension of what's currently implemented.  Cody also mentioned wanting to group by age range as determined by `NOW() - dob_column`, which we may also want to do for ICDS at some point.

Right now I'm soliciting feedback on how to proceed.

esoergel committed:
Test aggregate by date

esoergel committed:
Add support for conditional aggregations

esoergel committed:
Update sqlagg

esoergel commented on corehq/apps/userreports/tests/test_report_aggregation.py:
```diff
@@ -739,3 +757,89 @@ def test_with_prefilter(self):
+                        "age_at_registration between 7 and 12": "7-12",
+                    },
+                    'else_': '13+'
+                },
```
This test shows the functionality that's exposed.  The constrained version I'm considering would look something like this:
```JSON
{
    "type": "aggregate_range",
    "display": "age_range",
    "field": "age_at_registration",
    "column_id": "age_range",
    "ranges": {
        "0-6": [0, 6],
        "7-12": [7, 12],
    },
    "default": "13+"
}
```
That is, the user provides names, start and end points, and a default value, and that is transformed to the above.

To support similar use-cases, we'd need to write similar wrappers.

emord replied:
https://erikbern.com/2018/08/30/i-dont-want-to-learn-your-garbage-query-language.html

czue replied:
hahaha this is great.

> There should be a 30 year moratorium on inventing new query languages.

I've been thinking about this a lot (specifically in the context of UCR - which I think is a wonderful example of the pain of a custom query language). The thing I haven't figured out is how to get the benefits/reusability of a non-custom language without at least one of the following drawbacks:

- potentially exposing security holes by passing through raw queries
- doing a hell of a lot of work to build a giant custom parser
- supporting a bastardized subset of the query language that can be just as confusing as a custom one
- ~impossible to build a non-power-user UI on top of

the author suggests they just want to "download their data to sql" which is fine for the author but not really tenable in a SaaS product.

anyway, nothing actionable in this comment, but it resonates and also I don't quite know what the good alternative is.

czue replied:
I should add - generally I'm very much in favor of moving away from a custom query language unless we're willing to build a UI on top of it. The obvious choices for UCR would be to use `xpath` (better for our non-power users) or `sql` (better for our power-users and easier for our dev team). I also like the idea of transitioning more to SQL, though don't have my head wrapped around what the security precautions/rules we need to make are if we start to go down that route...

emord replied:
I had actually started something like this feature a while ago, but never quite got the interface that I was happy with. Reading that also really resonated with me and I think that putting SQL like this in is great. 

I don't believe this is a security hole as it's being escaped by sqlalchemy and I also don't believe it's anymore of a performance concern than UCR already is because it's not trying to query these columns, just performing calculations on the rows that would have already been returned

esoergel replied:
Got any thoughts on how/whether to document this?  Just say it's a SQL case expression and leave it at that?  Seems like it'd be best explained with examples.

czue replied:
to start you could add it to the readme with a link to [an example here](https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/userreports/examples/examples.md#data-source-indicators)

emord reviewed (approved):
:shipit:

emord committed:
SumWhenColumn

czue commented on corehq/apps/userreports/reports/specs.py:
```diff
@@ -352,6 +354,45 @@ def get_query_column_ids(self):
+        )
+
+    def get_whens(self):
+        transform_val = lambda val: "'{}'".format(val)
```
should this be called `quote_val`? also, what happens when there's a `'` in val? is there not a util in sqlalchemy that can safely do this?

czue reviewed (commented):
this is :fire:

esoergel committed:
Send bucket names as bindparams

This is how sqlalchemy.case wraps text normally, but sqlagg converts it to a
text() object, which allows for sql expressions, so string literals need to be
double quoted.  This approach is more constrained.

esoergel committed:
Merge pull request #21769 from dimagi/je/sum-when

SumWhenColumn
